File: custom_dir/custom_actions/loop_action.py
import json
import time

# -*- coding: UTF-8 -*-
from maa.context import Context
from maa.custom_action import CustomAction


from maa.agent.agent_server import AgentServer

@AgentServer.custom_action("LoopAction")
class LoopAction(CustomAction):
    def run(self,
            context: Context,
            argv: CustomAction.RunArg) -> bool:
        """
        :param argv: 运行参数, 包括action_list和loop_times
        :param context: 运行上下文
        :return: 是否执行成功
        """

        # 读取 custom_param 的参数：{"action_list": ["A", "B", "C"], "loop_times": x}
        json_data = json.loads(argv.custom_action_param)
        # 获取动作列表和循环次数
        action_list = json_data.get("action_list", [])
        loop_times = json_data.get("loop_times", 1)
        on_error= json_data.get("on_error", None)
        if not action_list or loop_times < 1:
            print("无效的action_list或loop_times")
            return False

        print(f"开始执行动作列表 {action_list}，循环 {loop_times} 次")

        for i in range(loop_times):
            print(f"第 {i + 1} 次循环开始")

            for action in action_list:
                action_time_name = f"第{i + 1}次{action}任务"
                print(f"执行动作: {action_time_name}")

                if on_error is not None:
                    context.run_task(action_time_name, {action_time_name:{"next":action,"timeout":1000,"on_error":on_error}})
                else:
                    context.run_task(action_time_name, {action_time_name:{"next":action,"timeout":1000}})

                time.sleep(0.5)  # 可根据需求设置不同的延迟

            print(f"第 {i + 1} 次循环结束")

        return True


    def stop(self) -> None:
        """停止执行的逻辑"""
        pass

File: agent/actions/loop_action/loop_action.go
package loop_action

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/MaaXYZ/maa-framework-go/v4"
)

type LoopAction struct{}

type LoopParams struct {
	ActionList []string `json:"action_list"`
	LoopTimes  int      `json:"loop_times"`
	OnError    string   `json:"on_error"`
}

// Run 执行循环动作
func (a *LoopAction) Run(ctx *maa.Context, arg *maa.CustomActionArg) bool {
	var params LoopParams
	if err := json.Unmarshal([]byte(arg.CustomActionParam), &params); err != nil {
		fmt.Printf("解析参数失败: %v\n", err)
		return false
	}

	if len(params.ActionList) == 0 || params.LoopTimes < 1 {
		fmt.Println("无效的action_list或loop_times")
		return false
	}

	fmt.Printf("开始执行动作列表 %v，循环 %d 次\n", params.ActionList, params.LoopTimes)

	for i := 0; i < params.LoopTimes; i++ {
		fmt.Printf("第 %d 次循环开始\n", i+1)

		for _, action := range params.ActionList {
			actionTimeName := fmt.Sprintf("第%d次%s任务", i+1, action)
			fmt.Printf("执行动作: %s\n", actionTimeName)

			taskOverride := map[string]any{
				actionTimeName: map[string]any{
					"next":    action,
					"timeout": 1000,
				},
			}

			if params.OnError != "" {
				taskOverride[actionTimeName].(map[string]any)["on_error"] = params.OnError
			}

			_, err := ctx.RunTask(actionTimeName, taskOverride)
			if err != nil {
				fmt.Printf("执行动作 %s 失败: %v\n", actionTimeName, err)
			}
			time.Sleep(500 * time.Millisecond)
		}

		fmt.Printf("第 %d 次循环结束\n", i+1)
	}

	return true
}